### PR TITLE
WIP: Shared object support

### PR DESCRIPTION
**Pull request by n-eiling**
Adds support for launching kernels from shared objects loaded during runtime using dlopen. As this is how pytorch uses CUDA, this should enable pytorch support to Cricket.
This involved adding support for decoding the fatbinary metadata before the embedded cubin ELF in binaries compiled by nvcc. Cricket becomes able to extract the cubin from a binary, send it via RPC to the server, where it will be executed using the driver APIs cuModuleLoadData.

- [x] decode fatbinary 
- [x] extract cubin 
- [x] send cubin to server
- [x] add registry for tranferred cubins and kernel functions so Cricket is able to identify them when launching kernels
- [x] switch over old kernel launching functionality to always use the new registry instead of relying on kernel locations being the same on client and server
- [x] use libelf to read kernel infos instead of relying on cuobjdump which does not support in-memory ELFs
- [x] read parameter infos using libelf.
- [x] enable reading CUDA elfs with debugging infos and compressed elfs
- [x] Test with minimal pytorch (deactivated some features, no kernel compression)
- [x] Test with default pytorch (no kernel compression)
- [x] Fix compression and test with default pytorch (with kernel compression)
- [x] fix CI
- [ ] large-scale test / verification
     - [ ] YOLOv5
- [x] cuDNN implementation

This also makes LD_PRELOADing on the server side not necessary anymore, as we now also extract and send cubins for normal applications.

This is work in progress. Addresses #6

### CODE REVIEW

**Comment by nravic** [comment]
Just saw this! Thanks for taking it on haha, was about to start this weekend. I'd love to help with this effort, let me know if there's anything I can do.

**Comment by jin-zhengnan** [comment]
@n-eiling  When will the entire test be completed, I am looking forward to this？

**Comment by n-eiling** [comment]
I updated my todo list. There are still some open issues that need adressing. CUDA relies on the .nv.info section for information regarding kernel parameter sizes and offsets. I used to parse them using cuobjdump, but this doesn't support in-memory ELFs - only files.

**Comment by n-eiling** [comment]
The last few days I was working on decompressing compressed CUDA fatbinaries. With default compiler parameters pytorch uses compressed CUDA code which we need to decompress in Cricket to be able to execute it. I worked on it out-of-tree (see https://github.com/n-eiling/cuda-fatbin-decompression) and will now start merging this into cricket.

**Comment by jin-zhengnan** [comment]
I have been following your project for a long time and have some questions that I would like to inquire about. Please help me answer them when you have time. Thank you very much. cudaRegisterVar function  why not directly call this functions on the server after client hijacking, but instead save them in server？ @n-eiling

**Comment by wangtianxia-sjtu** [comment]
I have noticed that you have marked the `Test with minimal pytorch` and `Test with default pytorch` as completed. How to reproduce these tests? I failed to run your code even with a simple pytorch script.

**Comment by n-eiling** [comment]
have a look here: https://github.com/RWTH-ACS/cricket/blob/share-object-support/docs/pytorch.md
note that this is still work in progress, so it probably will not work as you expect.
I used this pytorch script: https://github.com/RWTH-ACS/cricket/blob/share-object-support/tests/test_apps/pytorch_minimal.py

**Comment by jin-zhengnan** [comment]
padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling  and If the size is exactly divided by 8, do you still need to add padding?

**Comment by jin-zhengnan** [comment]
LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and  I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。

**Comment by n-eiling** [comment]
> padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling and If the size is exactly divided by 8, do you still need to add padding?

Isn't this was the original code achieves? Your suggestion doesn't work when the difference is exactly divisible by 8, while the original code does not add padding in this case.

> LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。

Thanks for catching the error! I fixed it. 
`cuGetExportTable` is part of the interface between runtime and driver API. I experimented a lot with getting the runtime API working while only implementing the driver API in Cricket. `cuGetExportTable` exchanges some pretty deep data structures between the APIs and I did not manage to figure out all the memory I need to copy. So it currently does not work correctly.

**Comment by jin-zhengnan** [comment]
> 



> > padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling and If the size is exactly divided by 8, do you still need to add padding?
> 
> Isn't this what the original code achieves? Your suggestion doesn't work when the difference is exactly divisible by 8, while the original code does not add padding in this case.
> 
> > LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。
> 
> Thanks for catching the error! I fixed it. `cuGetExportTable` is part of the interface between runtime and driver API. I experimented a lot with getting the runtime API working while only implementing the driver API in Cricket. `cuGetExportTable` exchanges some pretty deep data structures between the APIs and I did not manage to figure out all the memory I need to copy. So it currently does not work correctly.

Thanks for your answer, I think solve this is important, Maybe  This is a way  only  implementing the driver API,  if done , you can reduce the processing of many functions。
and help answer another question，cudaFuncGetAttributes exec fail in server，if use cuFuncGetAttribute replace is ok，you have some suggestion？

**Comment by n-eiling** [comment]
I created a separate issue for this (#21). Thanks for reporting the issue.

**Comment by wangtianxia-sjtu** [comment]
> have a look here: https://github.com/RWTH-ACS/cricket/blob/share-object-support/docs/pytorch.md note that this is still work in progress, so it probably will not work as you expect. I used this pytorch script: https://github.com/RWTH-ACS/cricket/blob/share-object-support/tests/test_apps/pytorch_minimal.py

Excuse me, but how to launch the cricket server? In your note, you only specify how to run the pytorch with cricket-client.so.

**Comment by wangtianxia-sjtu** [comment]
`cudaMemsetAsync` should be implemented because Pytorch may heavily use this api.

**Comment by wangtianxia-sjtu** [comment]
Do you have plans to support `nvrtc` (nv runtime compiling) APIs? 😊 Pytorch may use also use these APIs.

**Comment by n-eiling** [comment]
oh really? hm I haven't looked into nvrtc yet. Currently I am trying to get YOLOv5 working. Do you know some other useful pytorch examples I could use?

**Comment by wangtianxia-sjtu** [comment]
> oh really? hm I haven't looked into nvrtc yet. Currently I am trying to get YOLOv5 working. Do you know some other useful pytorch examples I could use?

Thanks for the reply. I just learn from a [paper](https://arxiv.org/pdf/2202.07848.pdf) from Microsoft where they claim that they intercept `nvrtcCompileProgram` call to extract kernel info.

I do not have any examples which will actually use `nvrtc` APIs. But I have found that in some versions of Pytorch, it will generate some kernels with `nvrtc` when you run `import torch`. And we can find in pytorch [source](https://github.com/pytorch/pytorch/blob/b9f81a483a7879cd3709fd26bcec5f1ee33577e6/aten/src/ATen/Context.h#L121) that it will use nvrtc.

**Comment by n-eiling** [comment]
I will merge this because the branch has diverged quite a bit and the original PR feature is working well. For pytorch, I still have some issues with the cudnnBackend API, which I will work on on a different branch.

**Comment by KangYingjie0** [comment]
> > padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling and If the size is exactly divided by 8, do you still need to add padding?
> 
> Isn't this what the original code achieves? Your suggestion doesn't work when the difference is exactly divisible by 8, while the original code does not add padding in this case.
> 
> > LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。
> 
> Thanks for catching the error! I fixed it. `cuGetExportTable` is part of the interface between runtime and driver API. I experimented a lot with getting the runtime API working while only implementing the driver API in Cricket. `cuGetExportTable` exchanges some pretty deep data structures between the APIs and I did not manage to figure out all the memory I need to copy. So it currently does not work correctly.

I tested the functionality of `cuGetExportTable`, it input char [16] by `exportTableId`, then output matched hidden function info (in libcuda.so) by `ppExportTable`. If `ppExportTable` is only a function pointer, the hidden function will be called by it.  I'm not entirely sure, `ppExportTable` could also contain more complex information.

**Comment by KangYingjie0** [comment]
> > padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling and If the size is exactly divided by 8, do you still need to add padding?
> 
> Isn't this what the original code achieves? Your suggestion doesn't work when the difference is exactly divisible by 8, while the original code does not add padding in this case.
> 
> > LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。
> 
> Thanks for catching the error! I fixed it. `cuGetExportTable` is part of the interface between runtime and driver API. I experimented a lot with getting the runtime API working while only implementing the driver API in Cricket. `cuGetExportTable` exchanges some pretty deep data structures between the APIs and I did not manage to figure out all the memory I need to copy. So it currently does not work correctly.

I tested the functionality of `cuGetExportTable`, it input char [16] by `exportTableId`, then output matched hidden function info (in libcuda.so) by `ppExportTable`. If `ppExportTable` is only a function pointer, the hidden function will be called by it.  I'm not entirely sure, `ppExportTable` could also contain more complex information.

**Comment by KangYingjie0** [comment]
> > padding = ((8 - (size_t)(input + input_read)) % 8); maybe change it padding = (8 - (size_t)(input + input_read)% 8); for this?@n-eiling and If the size is exactly divided by 8, do you still need to add padding?
> 
> Isn't this what the original code achieves? Your suggestion doesn't work when the difference is exactly divisible by 8, while the original code does not add padding in this case.
> 
> > LOGE(LOG_ERROR, "cannot find kernel %s kernel_info_t") this log miss the parameter？@n-eiling， and I want to know，cuGetExportTable is work normal in project？I find you do a lot of work for this。
> 
> Thanks for catching the error! I fixed it. `cuGetExportTable` is part of the interface between runtime and driver API. I experimented a lot with getting the runtime API working while only implementing the driver API in Cricket. `cuGetExportTable` exchanges some pretty deep data structures between the APIs and I did not manage to figure out all the memory I need to copy. So it currently does not work correctly.

I tested the functionality of `cuGetExportTable`, it input char [16] by `exportTableId`, then output matched hidden function info (in libcuda.so) by `ppExportTable`. If `ppExportTable` is only a function pointer, the hidden function will be called by it.  I'm not entirely sure, `ppExportTable` could also contain more complex information.
